crates/typst-pdf/src/font.rs

```diff
@@ -161,6 +163,67 @@ pub(crate) fn write_fonts(ctx: &mut PdfContext) {
 
         stream.finish();
     }
+
+    let emoji_font_map = ctx.emoji_font_map.take();
+    for font in emoji_font_map.items() {
+        for (font_index, subfont_id) in font.refs.iter().enumerate() {
+            let mut glyphs_to_instructions = BTreeMap::new();
+
+            let start = font_index * 256;
+            let end = (start + 256).min(font.emojis.len());
+            for (cid, emoji) in font.emojis[start..end].iter().enumerate() {
+                let page_ref = ctx.alloc.bump();
+                // create a fake page context for write_frame
+                // we are only interested in the contents of the page
+                let size = Axes::new(Abs::pt(1.0), Abs::pt(1.0));
+                let mut page_ctx = PageContext::new(ctx, page_ref, size);
+                page_ctx.bottom = size.y.to_f32();
+                page_ctx.transform(Transform {
+                    sx: Ratio::one(),
+                    ky: Ratio::zero(),
+                    kx: Ratio::zero(),
+                    sy: Ratio::new(-1.0),
+                    tx: Abs::zero(),
+                    ty: size.y,
+                });
+                page_ctx.content.start_color_glyph(emoji.image.width().to_f32());
+                write_frame(&mut page_ctx, &emoji.image);
+                let stream = page_ctx.content.finish();
+                ctx.pdf.stream(page_ref, &stream);
+
+                glyphs_to_instructions.insert(cid, page_ref);
+            }
+
+            let mut pdf_font = ctx.pdf.type3_font(*subfont_id);
+            pdf_font.bbox(font.bbox);
+            pdf_font.matrix([1.0, 0.0, 0.0, 1.0, 0.0, 0.0]);
+            {
+                let mut char_procs = pdf_font.char_procs();
+                for (gid, instructions_ref) in &glyphs_to_instructions {
+                    char_procs.pair(
+                        Name(eco_format!("glyph{gid}").as_bytes()),
+                        *instructions_ref,
+                    );
+                }
+            }
+            {
```

Comment:
Is there a good way to avoid these artificial scopes? I had to introduce them to not borrow `pdf_font` mutably twice.

Reply:
You can use [`.finish()`](https://docs.rs/pdf-writer/latest/pdf_writer/trait.Finish.html) for this.